### PR TITLE
BUG: Optimize: NewtonCG min crashes with xtol=0

### LINKED ISSUE #20214: BUG: minimize(method="newton-cg") crashes with UnboundLocalError if xtol=0
### Describe your issue.

I was trying to debug an issue where I was unable to find the minimum of a function with Newton-CG. As one of my debugging steps, I tried setting the solver option `xtol` to zero. I expected this to disable termination due to x convergence, but what actually happens is that it crashes.

### Reproducing Code Example

```python
import math
import scipy.optimize


def cosine(x):
    f = math.cos(x[0])
    print(f"f({x[0]}) = {f}")
    return f


def jac(x):
    ret = -math.sin(x)
    return ret


x0 = [0.1]
xtol = 0  # crashes with UnboundLocalError
# xtol = 1e-10  # works
result = scipy.optimize.minimize(cosine, x0=x0, jac=jac, method="newton-cg", options=dict(xtol=xtol))

print(result)
```

### Error message

```shell
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Input In [41], in <cell line: 19>()
     17 xtol = 0  # crashes with UnboundLocalError
     18 # xtol = 1e-10  # works
---> 19 result = scipy.optimize.minimize(cosine, x0=x0, jac=jac, method="newton-cg", options=dict(xtol=xtol))
     21 print(result)

File ~/.local/lib/python3.10/site-packages/scipy/optimize/_minimize.py:710, in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    708     res = _minimize_bfgs(fun, x0, args, jac, callback, **options)
    709 elif meth == 'newton-cg':
--> 710     res = _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback,
    711                              **options)
    712 elif meth == 'l-bfgs-b':
    713     res = _minimize_lbfgsb(fun, x0, args, jac, bounds,
    714                            callback=callback, **options)

File ~/.local/lib/python3.10/site-packages/scipy/optimize/_optimize.py:2233, in _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback, xtol, eps, maxiter, disp, return_all, c1, c2, **unknown_options)
   2230     update_l1norm = np.linalg.norm(update, ord=1)
   2232 else:
-> 2233     if np.isnan(old_fval) or np.isnan(update).any():
   2234         return terminate(3, _status_message['nan'])
   2236     msg = _status_message['success']

UnboundLocalError: local variable 'update' referenced before assignment
```

### SciPy/NumPy/Python version and system information

```shell
1.12.0 1.23.5 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.1
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.8
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-oj5r1eje/overlay/lib/python3.10/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

### PR DESCRIPTION
**Pull request by TimButters**
The minimize function crashes if `xtol` is set to `0`. This is because the initial value of a loop variable is `2 * xtol`, which is intended to ensure the loop runs at least once. However, if `xtol = 0` then the loop never runs, a variable isn't instantiated that is used outside of the loop, and an `UnboundLocalError` is raised.

To fix this, the initial value of the loop variable is set to the max float value.

#### Reference issue
Closes #20214 

#### What does this implement/fix?
UnboundLocalError is thrown if `xtol` is set to 0 for the `newton-cg` method in `minimize`.

### CODE REVIEW

**Review by lucascolley — APPROVED** [review]
LGTM!

**Review by tylerjereddy — COMMENTED** [review]
I'll let the optimize regulars chime in on whether a regression test is sensible to add, but this slightly modified version of the original reproducer fails before/passes after and seems fast:

```diff
--- a/scipy/optimize/tests/test_optimize.py
+++ b/scipy/optimize/tests/test_optimize.py
@@ -3120,3 +3120,19 @@ def test_sparse_hessian(method, sparse_type):
     assert res_dense.nfev == res_sparse.nfev
     assert res_dense.njev == res_sparse.njev
     assert res_dense.nhev == res_sparse.nhev
+
+
+def test_gh_20214():
+    def cosine(x):
+        f = np.cos(x[0])
+        return f
+
+    def jac(x):
+        ret = -np.sin(x)
+        return ret
+
+    result = optimize.minimize(cosine,
+                               x0=[0.1],
+                               jac=jac,
+                               method="newton-cg",
+                               options=dict(xtol=0, maxiter=2))
```

**Review by h-vetinari — DISMISSED** [review]
At least one error from #20214 occurs elsewhere, namely `update` is undefined in this branch:
https://github.com/scipy/scipy/blob/cd60e810c17442f71ab7f7788eccd1013fc265f1/scipy/optimize/_optimize.py#L2118-L2120

Looking at https://github.com/scipy/scipy/commit/c8c78d7171c074b8e5b97c57944b827fd13aa979, this commit failed to rename that instance of `update` to `update_l1norm`. CC @lorentzenchr 

We need a test that exercises this else-branch in any case. I haven't checked if @tylerjereddy's case does that, but to reach the `else:` clause of the `while:` loop, we must achieve `update_l1norm <= xtol` without having triggered any of the early `return` statements in the while body.

File: scipy/optimize/_optimize.py
Comment (h-vetinari) [inline scipy/optimize/_optimize.py:2028]: I cannot imagine this being the right solution to be honest. This means the `update_l1norm > xtol` clause of the while-loop will ~always be true, and completely ignores that this should have something to do with the initial norm.

The _very_ first thing here should be to fix the `UnboundLocalError`, not hide it from happening.

Comment (j-bowhay) [reply scipy/optimize/_optimize.py:2028]: I agree I think the correct fix is updating `update` -> `update_l1norm` on line 2119

Comment (TimButters) [reply scipy/optimize/_optimize.py:2028]: There are two related issues that I was trying to address with this:

1. UnboundLocalError if xtol=0
2. The function will not succeed if xtol=0, even if an exact solution can be found (as is the case with the example given in the bug (https://github.com/scipy/scipy/issues/20214)

I completely agree that the most important issue to fix is the Unbound Error, but if we only make the change `update` -> `update_l1norm` we will still fail to find the valid solution for the presented test case.

I could well be missing something, but I don't think the initial value _does_ need to be related to the initial norm. The variable is only used to enter the loop (which is guaranteed for any value that isn't zero). It is then updated at the end of the loop, it only appears to be used to track the exit condition of the while.

My idea with using the max float as the initial value is that it retains the guarantee of running at least once, and also stops any issues with people passing in extremely large values for `xtol` that could overflow.

I'd certainly welcome your thoughts @h-vetinari  and @j-bowhay - I'm very new to SciPy so could well be misunderstanding how this works.


Whatever the right solution, I should have added a test! I'm happy to add one in as suggested.

Comment (j-bowhay) [reply scipy/optimize/_optimize.py:2028]: Ok yes I see your point the initial assignment is effectively only to enter the loop (basically a do while). However I am not sure the change is correct because previously `2 * xtol` is always greater than `xtol` whereas now this not the case. Prehaps it should initially just be set to `np.inf`.

Regardless, the mistake on 2119 should be fixed and it would be good add the test.

Comment (j-bowhay) [reply scipy/optimize/_optimize.py:2028]: Ha sorry a very silly mistake from me, I miss read this as min not max, I think this should be fine

Comment (TimButters) [reply scipy/optimize/_optimize.py:2028]: Thanks. The latest revision also includes a test and corrects the mistake on line 2119, so I think that covers everything.

**Review by j-bowhay — APPROVED** [review]
Thanks @TimButters lgtm but I will wait to see if others are happy before merging

File: scipy/optimize/_optimize.py
Comment (h-vetinari) [reply scipy/optimize/_optimize.py:2028]: There's one thing that is being changed here, which is that previously, `update_l1norm` maintained the shape of `x0`, and in particular, `xtol` will _also_ have the shape of `x0`.

I don't think this ever worked for `len(x0)>1` though, as `while update_l1norm > xtol:` would then run into:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I don't know the code here well enough, but this stands out to me...

File: scipy/optimize/_optimize.py
Comment (h-vetinari) [inline scipy/optimize/_optimize.py:2119]: ... not least because you are also removing the call to `.any()` here.

**Review by h-vetinari — COMMENTED** [review]
I'm still unsure about the right initial value for `update_l1norm`, though I can see the argument that it's being updated at the end of the loop. Still, the interplay with the shape of `x` is a mystery to me. The code seems very brittle w.r.t. to putting in `x0=[1, 2]` or something like that.

Unless we can figure this out quickly _and_ comprehensively, I'd prefer an immediate fix for the unboundname error to go into 1.13, and the rest to be fixed for 1.14.

File: scipy/optimize/_optimize.py
Comment (lorentzenchr) [inline scipy/optimize/_optimize.py:2119]: Is it guaranteed that update_l1norm is nan whenever there‘s a nan in update?

File: scipy/optimize/tests/test_optimize.py
Comment (lorentzenchr) [inline scipy/optimize/tests/test_optimize.py:525]: Could we also test the value of the obtained solution?

**Review by lorentzenchr — APPROVED** [review]

**Comment by lorentzenchr** [comment]
Is there a test for the „nan-path“?

File: scipy/optimize/_optimize.py
Comment (TimButters) [reply scipy/optimize/_optimize.py:2119]: Yes, `np.linalg.norm` will always return `nan` if one or more elements of the vector is `nan`.

Comment (TimButters) [reply scipy/optimize/_optimize.py:2119]: Happy to add the `.any()` back in, but the norm is always going to be a scalar so it's unnecessary. The norm will always be `nan` if any of the elements of the vector (`update`) are `nan`.

I'm not sure about the shape side of this. It looks like previously `update_l1norm` was initialised to `2 * xtol`, and `xtol` is `len(x0) * avextol`. With `len(x0)` and `avextol` being scalars I think this has always just been a float value?

File: scipy/optimize/tests/test_optimize.py
Comment (TimButters) [reply scipy/optimize/tests/test_optimize.py:525]: Good point, I've added that to the test.

**Comment by TimButters** [comment]
> Is there a test for the „nan-path“?

@lorentzenchr yes, there is an existing test `test_nan_values` that checks this branch.
Line 1494 in `test_optimize.py`.

**Comment by TimButters** [comment]
@j-bowhay @lucascolley I think we have consensus now, would one of you be happy to merge this?

**Review by lucascolley — APPROVED** [review]
thanks @TimButters , I'll give it a few days for any more comments before merging.

As a note for the future, I'd recommend that you leave resolving conversations to maintainers - a thorough reviewer will probably want to check any conversations that were resolved by the author, so it is easier to leave them as is.

**Comment by TimButters** [comment]
Thanks for the feedback @lucascolley, I thought I was doing the right thing there. I’ll leave them open in the future.

**Comment by lucascolley** [comment]
> I'll give it a few days for any more comments before merging.

Here we are, thanks @TimButters !

**Comment by j-bowhay** [comment]
Thanks @TimButters a great couple of first contributions, keep them coming!

**Pull request merged**
